staker/challenge-cache/cache.go

```diff
@@ -0,0 +1,242 @@
+			hashes.bin
+			subchallenge-level-1-big-step-100/
+				hashes.bin
+
+We namespace top-level block challenges by wavm module root. Then, we can retrieve
+the hashes for any data within a challenge or associated subchallenge based on the hierarchy above.
+*/
+
+package challengecache
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	ErrNotFoundInCache    = errors.New("no found in challenge cache")
+	ErrFileAlreadyExists  = errors.New("file already exists")
+	ErrNoHashes           = errors.New("no hashes being written")
+	hashesFileName        = "hashes.bin"
+	wavmModuleRootPrefix  = "wavm-module-root"
+	rollupBlockHashPrefix = "rollup-block-hash"
+	messageNumberPrefix   = "message-num"
+	bigStepPrefix         = "big-step"
+	challengeLevelPrefix  = "subchallenge-level"
+)
+
+// HistoryCommitmentCacher can retrieve history commitment hashes given lookup keys.
+type HistoryCommitmentCacher interface {
+	Get(lookup *Key, numToRead uint64) ([]common.Hash, error)
+	Put(lookup *Key, hashes []common.Hash) error
+}
+
+// Cache for history commitments on disk.
+type Cache struct {
+	baseDir string
+}
+
+// New cache from a base directory path.
+func New(baseDir string) (*Cache, error) {
+	if _, err := os.Stat(baseDir); err != nil {
+		if err := os.MkdirAll(baseDir, os.ModePerm); err != nil {
+			return nil, fmt.Errorf("could not make base cache directory %s: %w", baseDir, err)
+		}
+	}
+	return &Cache{
+		baseDir: baseDir,
+	}, nil
+}
+
+// Key for cache lookups includes the wavm module root of a challenge, as well
+// as the heights for messages and big steps as needed.
+type Key struct {
+	RollupBlockHash common.Hash
+	WavmModuleRoot  common.Hash
+	MessageHeight   uint64
+	StepHeights     []uint64
+}
+
+// Get a list of hashes from the cache from index 0 up to a certain index. Hashes are saved as files in the directory
+// hierarchy for the cache. If a file is not present, ErrNotFoundInCache
+// is returned.
+func (c *Cache) Get(
+	lookup *Key,
+	numToRead uint64,
+) ([]common.Hash, error) {
+	fName, err := determineFilePath(c.baseDir, lookup)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := os.Stat(fName); err != nil {
+		log.Warn("Cache miss", "fileName", fName)
+		return nil, ErrNotFoundInCache
+	}
+	log.Debug("Cache hit", "fileName", fName)
+	f, err := os.Open(fName)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Error("Could not close file after reading", "err", err, "file", fName)
+		}
+	}()
+	return readHashes(f, numToRead)
+}
+
+// Put a list of hashes into the cache.
+// Hashes are saved as files in a directory hierarchy for the cache.
+// This function first creates a temporary file, writes the hashes to it, and then renames the file
+// to the final directory to ensure atomic writes.
+func (c *Cache) Put(lookup *Key, hashes []common.Hash) error {
```

Comment:
I like the tempfile approach, but there is some overcomplication here and in the directory structure.

I suggest:

during New/Initialize:


* create a single basedir/temp directory. Check that you can MkdirTemp a subdir of temp and delete it. Also - if "temp" dir exists and is not empty - issue a log.Warning.
* always create tempdirs as a subdir ot "temp" that way it's always easy to clean
* create tempfile as a single file directly under the tempdir. the path doesn't really help here.
* after move succeeds - delete the tempdir